### PR TITLE
CP-18825: Remove FIST and ensure signature must be verified.

### PR DESCRIPTION
**Pull request by huizh**
Signed-off-by: Hui Zhang hui.zhang@citrix.com

### CODE REVIEW

File: scripts/extensions/pool_update.precheck [hunk 1]
```diff
@@ -49,10 +48,10 @@ def parse_control_package(session, yum_url):
     return items[0].getAttribute('control')
 
 
-def execute_precheck(session, control_package, yum_conf_file, nogpgcheck):
+def execute_precheck(session, control_package, yum_conf_file):
     FNULL = open(os.devnull, 'w')
     retcode = subprocess.call('yum -c ' + yum_conf_file + ' install -y ' +
```
Comment (lindig) [inline scripts/extensions/pool_update.precheck]: Just a general comment: I think it is a bad idea to create a string that is interpreted by a shell unless this is absolutely necessary and here it doesn't look like this is the case. It is much safer to provide arguments as an array of strings that are not passed through a shell. See: https://docs.python.org/2/library/subprocess.html#using-the-subprocess-module

**Comment by phuslu** [comment]
@lindig Thanks. I updated the subprocess call in 2nd commit https://github.com/xenserver/xen-api/pull/166/commits/6b29db5c188acb0519fca69a8e3b53ef710ecb4c

**Comment by ghost** [comment]
Thanks @lindig I updated the subprocess call in 2nd commit https://github.com/xenserver/xen-api/pull/166/commits/6b29db5c188acb0519fca69a8e3b53ef710ecb4c

**Comment by lindig** [comment]
Thanks. I'd like somebody with Python experience to take a look, too. @mseri?

File: scripts/extensions/pool_update.apply [hunk 2]
```diff
@@ -36,16 +36,25 @@ def failure_message(code, params):
...
-                              nogpgcheck + update_package, shell=True, stdout=FNULL, stderr=FNULL)
-    if retcode != 0:
-        raise ApplyFailure(
-            'Failed to install update_package (%r).' % update_package)
+def execute_apply(session, update_package, yum_conf_file):
+    cmd = ['yum', 'install', '-y', '--noplugins', '-c', yum_conf_file, update_package]
+    p = subprocess.Popen(cmd, shell=False, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, close_fds=True)
+    p.wait()
+    output = p.stdout.read()
+    xcp.logger.info('pool_update.apply %r returncode=%r output=%r', cmd, p.returncode, output)
```
Comment (mseri) [inline scripts/extensions/pool_update.apply]: Are we sure that the output here is never huge?

File: scripts/extensions/pool_update.apply [hunk 3]
```diff
@@ -36,16 +36,25 @@ def failure_message(code, params):
...
+    if p.returncode != 0:
+        lines = output.splitlines()
+        errorlines = [l for l in lines if l.startswith('Error: ')]
+        if errorlines:
+            errmsg = errorlines[-1].split(':', 1)[-1].strip()
+        else:
+            errmsg = lines[-1]
```
Comment (mseri) [inline scripts/extensions/pool_update.apply:51]: Are usually the errors starting with `Error:`? I was making some experiments and `lines` on my machine looks like `['Loaded plugins: fastestmirror', 'There are no enabled repos.', ' Run "yum repolist all" to see the repos you have.', ' You can enable repos with yum-config-manager --enable <repo>']`. Is there something specific in `yum_conf_file` that makes the output show the prefix?

File: scripts/extensions/pool_update.apply (hunk 2 above)
Comment (mseri) [inline scripts/extensions/pool_update.apply:44]: Are we sure that `output` here is not a huge string?

File: scripts/extensions/pool_update.precheck [hunk 4]
```diff
@@ -49,16 +50,25 @@ def parse_control_package(session, yum_url):
...
-    FNULL = open(os.devnull, 'w')
-    retcode = subprocess.call('yum -c ' + yum_conf_file + ' install -y ' +
-                              nogpgcheck + control_package, shell=True, stdout=FNULL, stderr=FNULL)
-    if retcode != 0:
+def execute_precheck(session, control_package, yum_conf_file):
+    cmd = ['yum', 'install', '-y', '--noplugins', '-c', yum_conf_file, control_package]
+    p = subprocess.Popen(cmd, shell=False, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, close_fds=True)
+    p.wait()
+    output = p.stdout.read()
+    xcp.logger.info('pool_update.precheck %r returncode=%r output=%r', cmd, p.returncode, output)
```
Comment (mseri) [inline scripts/extensions/pool_update.precheck:58]: As above, are we sure that `output` here is not a huge string?

Comment (lindig) [reply scripts/extensions/pool_update.precheck:58]: I'm not sure but if this is for updating RPMs that we control the output of `yum` should not reach a size where we need to worry about this. I would be surprised if this is even above 10k.

File: scripts/extensions/pool_update.apply (hunk 3 above)
Comment (ghost) [reply scripts/extensions/pool_update.apply:51]: For `yum install` errors, usually the error prefix is 'Error: ', we could confirm `grep 'Error: '` in https://github.com/rpm-software-management/yum
(We didn't add something specific in `yum_conf_file`)

File: scripts/extensions/pool_update.precheck (hunk 4 above)
Comment (ghost) [reply scripts/extensions/pool_update.precheck:58]: Thanks @lindig , For the is case(`yum install` a update_package), I also think its output should not big.

**Comment by lindig** [comment]
I'd be fine with merging this.

File: scripts/extensions/pool_update.precheck (hunk 1 above)
Comment (srowe) [reply scripts/extensions/pool_update.precheck]: I was going to submit a separate PR for this.

**Comment by mseri** [comment]
Fine with me as well

File: scripts/extensions/pool_update.apply [hunk 5]
```diff
@@ -36,16 +36,25 @@ def failure_message(code, params):
...
-def execute_apply(session, update_package, yum_conf_file, nogpgcheck):
-    FNULL = open(os.devnull, 'w')
-    retcode = subprocess.call('yum -c ' + yum_conf_file + ' install -y ' +
-                              nogpgcheck + update_package, shell=True, stdout=FNULL, stderr=FNULL)
-    if retcode != 0:
-        raise ApplyFailure(
-            'Failed to install update_package (%r).' % update_package)
+def execute_apply(session, update_package, yum_conf_file):
+    cmd = ['yum', 'install', '-y', '--noplugins', '-c', yum_conf_file, update_package]
+    p = subprocess.Popen(cmd, shell=False, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, close_fds=True)
```
Comment (srowe) [inline scripts/extensions/pool_update.apply:41]: `shell=False` is the default, there's no need to include it.

File: scripts/extensions/pool_update.apply (hunk 2 above)
Comment (srowe) [reply scripts/extensions/pool_update.apply:44]: These two lines would be better as

```
output, _ = p.communicate()
```

Comment (mseri) [reply scripts/extensions/pool_update.apply:44]: If we use `p.communicate()`, why don't we set `stdout=None` (default) and `stderr=supprocess.PIPE` and then `_, output = p.communicate()`? It should also make sure that the output is small, if any.

**Pull request merged**